helix-term/src/commands.rs

```diff
@@ -4533,18 +4535,118 @@ pub fn completion(cx: &mut Context) {
 }
 
 // comments
-fn toggle_comments(cx: &mut Context) {
+type CommentTransactionFn = fn(
+    line_token: Option<&str>,
+    block_tokens: Option<&[BlockCommentToken]>,
+    doc: &Rope,
+    selection: &Selection,
+) -> Transaction;
+
+fn toggle_comments_impl(cx: &mut Context, comment_transaction: CommentTransactionFn) {
     let (view, doc) = current!(cx.editor);
-    let token = doc
+    let line_token: Option<&str> = doc
+        .language_config()
+        .and_then(|lc| lc.comment_tokens.as_ref())
+        .and_then(|tc| tc.first())
+        .map(|tc| tc.as_str());
+    let block_tokens: Option<&[BlockCommentToken]> = doc
         .language_config()
-        .and_then(|lc| lc.comment_token.as_ref())
-        .map(|tc| tc.as_ref());
-    let transaction = comment::toggle_line_comments(doc.text(), doc.selection(view.id), token);
+        .and_then(|lc| lc.block_comment_tokens.as_ref())
+        .map(|tc| &tc[..]);
+
+    let transaction =
+        comment_transaction(line_token, block_tokens, doc.text(), doc.selection(view.id));
 
     doc.apply(&transaction, view.id);
     exit_select_mode(cx);
 }
 
+fn toggle_comments(cx: &mut Context) {
```

Comment:
This needs some documentation on how it works. If I'm commenting a fresh set of lines, does it use block or line comments? I'd prefer it to use line comments if it can, because it makes it easier to partially uncomment

Reply:
Also, let's say that I'm trying to uncomment two lines with line comments, I could start my selection in the middle of the first line, then end in the middle of the second line and both lines would uncomment (in other words I don't have to do a perfect selection). Is that behaviour retained?

Reply:
ah yes thats a good idea, I'll do that now, but the idea of this is it doesn't change anything at all if there is a line comment, if the whole selection is surrounded in block comment then it will uncomment that region, if the whole selection is line commented the uncomment that, if there is no comments or not all lines are commented with block comments and there is only block comment tokens it will comment each non commented line individually with those tokens, if each line is commented with line comments then it will un-comment all the lines.

Reply:
this is how I just commented it:
```rust
/// commenting behavior:
/// 1. only line comment tokens -> line comment
/// 2. each line block commented -> uncomment all lines
/// 3. whole selection block commented -> uncomment selection
/// 4. all lines not commented and block tokens -> comment uncommented lines
/// 5. no comment tokens and not block commented -> line comment
```